### PR TITLE
refactor node-disk-manager code

### PR DESCRIPTION
**Pull request by shovanmaity**
### This pr is for refactoring node-disk-manager code. ###
Why this changes required?
```
- Introduce probe concept in node-disk-manager and refactor code. 
```

Restructured directory structure is given bellow -
- previous cmd/controller
```
├── controller
│   ├── ctrl.go
│   ├── ctrl_main.go
│   └── ctrl_test.go
```
- updated cmd/controller
```
├── controller
│   ├── controller.go
│   ├── disk.go
│   ├── diskstore.go
│   └── probe.go
├── probe
│   ├── eventhandler.go
│   ├── probe.go
│   └── udevprobe.go
```
- previous pkg/udev
```
├── udev
│   └── udev.go
```
- updated pkg/udev
```
├── udev
│   ├── common.go
│   ├── device.go
│   ├── enumerate.go
│   ├── listentry.go
│   └── udev.go
```
TODO - write ut for probe

### CODE REVIEW

File: cmd/app/command/start.go [hunk 1]
```diff
@@ -46,7 +47,8 @@ func NewCmdStart() *cobra.Command {
 			metrics.StartingTime = time.Now()
 			// Start HTTP server for /metrics endpoint
 			go server.StartHttpServer()
-			controller.Watch(options.kubeconfig)
+			probe.Start()
```
Comment (kmova) [inline cmd/app/command/start.go]: Does this call the probeInterface Start() ? If yes, why is this required?

File: cmd/controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	NDMVersion  = "openebs.io/v1alpha1"    // NDMVersion is the CR version.
+	NDMHostKey  = "kubernetes.io/hostname" // NDMHostKey is host name label prefix.
+	NDMActive   = "Active"                 // NDMActive is constant for active resource status.
+	NDMInactive = "Inactive"               // NDMInactive is constant for inactive resource status.
+)
+
+// ProbeChannel is used to send a copy of controller object to each probe.
+// Each probe can get the copy of controller struct any time they
+// need to read the channel.
+var ProbeChannel = make(chan *Controller)
```
Comment (kmova) [inline cmd/controller/controller.go]: Should we rename this to "ControllerBroadcastChannel" -- since this is mainly used for advertising the controller object to any of the listeners/receivers. It could be "probes" today and could be others tomorrow..

File: cmd/controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	controller, err := newController(kuberconfig)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	if err = controller.run(2, stopCh); err != nil {
+		glog.Fatalf("error running controller: %s", err.Error())
+	}
+}
+
+// run will discover the local disks and push them to the etcd.
```
Comment (kmova) [inline cmd/controller/controller.go]: It appears that this comment needs to be updated with the refactored code.

File: cmd/controller/controller.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	<-stopCh
+	glog.Info("shutting down the controller")
+	return nil
+}
+
+// DeviceList queries all the disk resources for this host from etcd and
+// prints them to the standard output. This function is called when we execute
+// cli command ndm device list.
+func DeviceList(kuberconfig string) {
+	controller, err := newController(kuberconfig)
```
Comment (kmova) [inline cmd/controller/controller.go]: should this and the following functions be moved to diskstor?

File: cmd/probe/eventhandler.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+package probe
+
+import (
+	"github.com/golang/glog"
+	"github.com/openebs/node-disk-manager/cmd/controller"
+)
+
+// UdevEventMessage channel used to send event message
+var UdevEventMessage = make(chan EventMessage)
```
Comment (kmova) [inline cmd/probe/eventhandler.go]: Can this be renamed? Also, is this going to be only for udev event message? Or can this also be used for other probe event messages?

File: pkg/apis/ndm/register.go [hunk 6]
```diff
@@ -1,5 +1,5 @@
 package ndm
 
 const (
-	GroupName = "openebs.io"
+	GroupName = "ndm"
```
Comment (kmova) [inline pkg/apis/ndm/register.go]: what is this change for?

Comment (shovanmaity) [reply pkg/apis/ndm/register.go]: we will discuss on it. directory structure should be like `pkg/apis/openebs.io/` as the groupName is `openebs.io`

File: cmd/app/command/start.go (hunk 1 above)
Comment (shovanmaity) [reply cmd/app/command/start.go]: This is not calling `probeInterface Start()` directly this is required to invoke `init()` of each probe which calls `registerProbe()` and in `registerProbe()` `start()` is called if probe status is enabled.

File: cmd/controller/controller.go (hunk 2 above)
Comment (shovanmaity) [reply cmd/controller/controller.go]: Renamed the channel as `ControllerBroadcastChannel`

File: cmd/controller/controller.go (hunk 3 above)
Comment (shovanmaity) [reply cmd/controller/controller.go]: Updated comment added.

File: cmd/controller/controller.go (hunk 4 above)
Comment (shovanmaity) [reply cmd/controller/controller.go]: `diskstor.go` contains crate/update/delete/deactivate of disk .. is it ok to put `DeviceList()` there. as this function is used by cli command `ndm device list`

File: cmd/probe/eventhandler.go (hunk 5 above)
Comment (shovanmaity) [reply cmd/probe/eventhandler.go]: `udevprobe.listen()` uses this channel to get event and later udev monitor will use to send event details to udev probe. so this channel is udev specific. I will put it in udevprobe.go file.

File: cmd/controller/probe.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+func (ps sortableProbes) Less(i, j int) bool {
+	return ps[i].Priority < ps[j].Priority
+}
+
+// AddNewProbe adds new probe to controller object
+func (c *Controller) AddNewProbe(probe *Probe) {
+	probes := c.Probes
+	probes = append(probes, probe)
+	sort.Sort(sortableProbes(probes))
+	c.Probes = probes
```
Comment (kmova) [inline cmd/controller/probe.go]: should c.Probes be lock protected?

File: cmd/probe/udevprobe.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+func init() {
+	go func() {
+		ctrl := <-controller.ControllerBroadcastChannel
+		if ctrl == nil {
+			glog.Error("unable to configure", udevProbeName)
+			return
+		}
+		udevprobe := newUdevProbe(ctrl)
+		var pi controller.ProbeInterface = udevprobe
```
Comment (kmova) [inline cmd/probe/udevprobe.go]: Is this variable declaration required?

File: cmd/controller/probe.go (hunk 7 above)
Comment (shovanmaity) [reply cmd/controller/probe.go]: Yes, good point we can use one lock in controller, as all the probes will try to register in go routines. I will add that.

File: cmd/probe/udevprobe.go (hunk 8 above)
Comment (shovanmaity) [reply cmd/probe/udevprobe.go]: No , changed that.

File: cmd/controller/probe.go (hunk 7 above)
Comment (shovanmaity) [reply cmd/controller/probe.go]: done.

File: cmd/controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+		Clientset:     crdClient,
+		Probes:        probes,
+		Mutex:         mutex,
+	}
+	return controller, nil
+}
+
+// Start scans the locally attached disks and push them to etcd.
+// This function is called when we execute cli command ndm start.
+func Start(kuberconfig string) {
```
Comment (sagarkrsd) [inline cmd/controller/controller.go]: Should not it be kubeconfig ?

File: cmd/controller/controller.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+type Controller struct {
+	HostName      string               // host name is host name in which disk is attached
+	KubeClientset kubernetes.Interface // kubeclientset is standard kubernetes clientset
+	Clientset     clientset.Interface  // sampleclientset is clientset for our own API group
+	Probes        []*Probe             // Probes are the registered probes like udev/smart.
+	Mutex         *sync.Mutex          // Mutex is used to lock and unlock Controller
+}
+
+// newController returns a controller pointer for any error case it will return nil
+func newController(kuberconfig string) (*Controller, error) {
```
Comment (sagarkrsd) [inline cmd/controller/controller.go]: Shouldn't it be kubeconfig ?

File: cmd/controller/controller.go [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+// ControllerBroadcastChannel is used to send a copy of controller object to each probe.
+// Each probe can get the copy of controller struct any time they need to read the channel.
+var ControllerBroadcastChannel = make(chan *Controller)
+
+// Controller is the controller implementation for do resources
+type Controller struct {
+	HostName      string               // host name is host name in which disk is attached
+	KubeClientset kubernetes.Interface // kubeclientset is standard kubernetes clientset
+	Clientset     clientset.Interface  // sampleclientset is clientset for our own API group
```
Comment (a4abhishek) [inline cmd/controller/controller.go]: `sampleclientset` or just `Clientset`?

File: cmd/controller/controller.go (hunk 9 above)
Comment (shovanmaity) [reply cmd/controller/controller.go]: yes.

File: cmd/controller/controller.go (hunk 10 above)
Comment (shovanmaity) [reply cmd/controller/controller.go]: yes.

**Comment by shovanmaity** [comment]
@a4abhishek addressed your comments please check it once.

**Comment by a4abhishek** [comment]
> @sagarkrsd , @a4abhishek, addressed your comments please check it once.

Yes @shovanmaity looks good.

File: cmd/app/command/start.go (hunk 1 above)
Comment (kmova) [reply cmd/app/command/start.go]: Can you put a comment above this call similar to StartHttpServer, on what proble.Start will accomplish.

File: cmd/controller/disk.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+	Capacity         uint64          // Capacity is capacity of a disk
+	Model            string          // Model is model no of a disk
+	Serial           string          // Serial is serial no of a disk
+	Vendor           string          // Vendor is vendor of a disk
+	Path             string          // Path is dev path of a disk like /dev/sda
+}
+
+// ProbeIdentifiers contains some keys to uniquely identify each disk by probe
+type ProbeIdentifier struct {
+	Uuid           string // Uuid is uid of disk
```
Comment (kmova) [inline cmd/controller/disk.go]: How is this different from the Uuid in the above DiskInfo structure?

File: cmd/controller/disk.go [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+*/
+
+package controller
+
+import apis "github.com/openebs/node-disk-manager/pkg/apis/openebs.io/v1alpha1"
+
+// DiskInfo contains details of a disk which can be converted into api.Disk
+type DiskInfo struct {
+	ProbeIdentifiers ProbeIdentifier // ProbeIdentifiers contains some keys to uniquely identify each disk by probe
+	HostName         string          // HostName is hostName in which disk is attached
```
Comment (kmova) [inline cmd/controller/disk.go]: HostName contains the node's hostname in which this disk is attached.

File: cmd/probe/eventhandler.go [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+	Devices []*controller.DiskInfo // list of disks details
+}
+
+// addDiskEvent fill disk details from different probes and push it to etcd
+func (pe *ProbeEvent) addDiskEvent(msg EventMessage) {
+	for _, diskDetails := range msg.Devices {
+		func() {
+			glog.Info("processing data for ", diskDetails.ProbeIdentifiers.Uuid)
+			diskDetails.HostName = pe.Controller.HostName
+			diskDetails.Uuid = diskDetails.ProbeIdentifiers.Uuid
```
Comment (kmova) [inline cmd/probe/eventhandler.go]: As per the diskDetails -- this should be an generated uuid by the ndm. Is there some change here?

File: cmd/controller/disk.go (hunk 12 above)
Comment (shovanmaity) [reply cmd/controller/disk.go]: If we want to change disk uid generation functionality then we need this. Earlier @pawanpraka1  told we may revisit this uid generation function. As main disk discover and attach/detach event is generated by udev probe only at the time of discovering or event it generate one uid and that uid is same as disk uid that's why both are same.

File: cmd/probe/eventhandler.go (hunk 14 above)
Comment (shovanmaity) [reply cmd/probe/eventhandler.go]: If we want to change disk uid generation functionality then we need this. Earlier @pawanpraka1  told we may revisit this uid generation function. As main disk discover and attach/detach event is generated by udev probe only at the time of discovering or event it generate one uid and that uid is same as disk uid that's why both are same.

File: pkg/udev/common.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+		Serial: device.UdevDeviceGetPropertyValue(UDEV_SERIAL),
+		Vendor: device.UdevDeviceGetPropertyValue(UDEV_VENDOR),
+		Path:   device.UdevDeviceGetPropertyValue(UDEV_DEVNAME),
+		Size:   size,
+	}
+	return diskDetails
+}
+
+// GetUid returns unique id for the disk block device
+func (device *UdevDevice) GetUid() string {
```
Comment (shovanmaity) [inline pkg/udev/common.go]: @kmova here we are udev probe is generating disk uuid hash.

File: cmd/controller/disk.go (hunk 13 above)
Comment (shovanmaity) [reply cmd/controller/disk.go]: done.

File: cmd/controller/controller.go [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+	controller, err := newController(kubeconfig)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	if err = controller.run(2, stopCh); err != nil {
+		glog.Fatalf("error running controller: %s", err.Error())
+	}
+}
+
+// run broadcasts controller copy to each probe.
```
[the remark this answers is not in the document]
Comment (shovanmaity) [reply cmd/controller/controller.go]: comment added.

File: cmd/controller/controller.go [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+	HostName      string               // HostName is host name in which disk is attached
+	KubeClientset kubernetes.Interface // KubeClientset is standard kubernetes clientset
+	Clientset     clientset.Interface  // Clientset is clientset for our own API group
+	Probes        []*Probe             // Probes are the registered probes like udev/smart
+	Mutex         *sync.Mutex          // Mutex is used to lock and unlock Controller
+}
+
+// newController returns a controller pointer for any error case it will return nil
+func newController(kubeconfig string) (*Controller, error) {
+	mutex := &sync.Mutex{}
```
[the remark this answers is not in the document]
Comment (shovanmaity) [reply cmd/controller/controller.go]: changed.

**Comment by shovanmaity** [comment]
@kmova as we discussed I have tested it in local setup and 2 node gke cluster.
node-disk-manager pod
```
shovan_maity@cloudshell:~ (mayaiodev-attur)$ kubectl get pod
NAME                      READY     STATUS    RESTARTS   AGE
node-disk-manager-28gnk   1/1       Running   0          17s
node-disk-manager-f8sjw   1/1       Running   0          17s
```
disk resources
```
shovan_maity@cloudshell:~ (mayaiodev-attur)$ kubectl get disk
NAME                                    AGE
disk-ecf109ad98be73104b0c9c77924e47b3   27s
disk-fbb19aa12f332c8b69baadb8626f6df8   27s
```
log of node-disk-manager-28gnk
```
shovan_maity@cloudshell:~ (mayaiodev-attur)$ kubectl logs -f node-disk-manager-28gnk
I0702 09:27:12.896324       1 probe.go:54] starting probe
I0702 09:27:12.898026       1 controller.go:113] started the controller
I0702 09:27:12.898322       1 probe.go:76] configured udev probe
I0702 09:27:12.900970       1 server.go:39] Starting HTTP server at http://localhost:9090/metrics for metrics.
I0702 09:27:12.901230       1 udevprobe.go:173] starting udev probe listener
I0702 09:27:12.917474       1 eventhandler.go:47] processing data for disk-ecf109ad98be73104b0c9c77924e47b3
I0702 09:27:12.917611       1 eventhandler.go:52] disk details filled by udev probe
I0702 09:27:12.930226       1 diskstore.go:39] create failed, trying to update the disk object : disks.openebs.io "disk-ecf109ad98be73104b0c9c77924e47b3" already exists
I0702 09:27:12.937528       1 diskstore.go:75] updated disk object : disk-ecf109ad98be73104b0c9c77924e47b3
```
log of node-disk-manager-f8sjw
```
shovan_maity@cloudshell:~ (mayaiodev-attur)$ kubectl logs -f node-disk-manager-f8sjw
I0702 09:27:12.885829       1 probe.go:54] starting probe
I0702 09:27:12.887819       1 controller.go:113] started the controller
I0702 09:27:12.888207       1 probe.go:76] configured udev probe
I0702 09:27:12.891299       1 server.go:39] Starting HTTP server at http://localhost:9090/metrics for metrics.
I0702 09:27:12.891629       1 udevprobe.go:173] starting udev probe listener
I0702 09:27:12.914728       1 eventhandler.go:47] processing data for disk-ecf109ad98be73104b0c9c77924e47b3
I0702 09:27:12.914893       1 eventhandler.go:52] disk details filled by udev probe
I0702 09:27:12.923571       1 diskstore.go:30] created disk object : disk-ecf109ad98be73104b0c9c77924e47b3
I0702 09:27:12.923718       1 eventhandler.go:47] processing data for disk-fbb19aa12f332c8b69baadb8626f6df8
I0702 09:27:12.923795       1 eventhandler.go:52] disk details filled by udev probe
I0702 09:27:12.935248       1 diskstore.go:30] created disk object : disk-fbb19aa12f332c8b69baadb8626f6df8
```
Describe disk
```
shovan_maity@cloudshell:~ (mayaiodev-attur)$ kubectl describe disk disk-ecf109ad98be73104b0c9c77924e47b3
Name:         disk-ecf109ad98be73104b0c9c77924e47b3
Namespace:
Labels:       kubernetes.io/hostname=gke-csi-cluster-default-pool-afae5527-06m2
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Kind:         Disk
Metadata:
  Cluster Name:
  Creation Timestamp:  2018-07-02T09:27:12Z
  Generation:          1
  Resource Version:    2856245
  Self Link:           /apis/openebs.io/v1alpha1/disks/disk-ecf109ad98be73104b0c9c77924e47b3
  UID:                 1a1c2d75-7dda-11e8-b492-42010a80009d
Spec:
  Capacity:
    Storage:  107374182400
  Details:
    Model:   PersistentDisk
    Serial:  persistent-disk-0
    Vendor:  Google
  Path:      /dev/sda
Status:
  State:  Active
Events:   <none>
shovan_maity@cloudshell:~ (mayaiodev-attur)$ kubectl describe disk disk-fbb19aa12f332c8b69baadb8626f6df8
Name:         disk-fbb19aa12f332c8b69baadb8626f6df8
Namespace:
Labels:       kubernetes.io/hostname=gke-csi-cluster-default-pool-afae5527-r11x
Annotations:  <none>
API Version:  openebs.io/v1alpha1
Kind:         Disk
Metadata:
  Cluster Name:
  Creation Timestamp:  2018-07-02T09:27:12Z
  Generation:          1
  Resource Version:    2856244
  Self Link:           /apis/openebs.io/v1alpha1/disks/disk-fbb19aa12f332c8b69baadb8626f6df8
  UID:                 1a1dd4a8-7dda-11e8-b492-42010a80009d
Spec:
  Capacity:
    Storage:  11811160064
  Details:
    Model:   PersistentDisk
    Serial:  ganesh-disk1
    Vendor:  Google
  Path:      /dev/sdb
Status:
  State:  Active
Events:   <none>
```

File: cmd/app/command/start.go (hunk 1 above)
Comment (shovanmaity) [reply cmd/app/command/start.go]: yes, added.
